### ISSUE TITLE
FRAGMOS - RollConventionEnum is missing roll day "31"

### ISSUE THREAD

**Issue by JBZ-Fragmos** [finos/common-domain-model#2330]
**PROPOSAL**
add **"31"** in the Enum list of **roll days**, that currently ends at "30"

**BUSINESS NEED**
source of issue : actual termsheets from Client in which roll day of the IRS is  31 and we could not represent it in CDM...

![image](https://github.com/finos/common-domain-model/assets/92978493/378345dc-825b-479b-bd0e-7b6b0dba58da)

**Comment by iansloyan** [comment]
@JBZ-Fragmos I love the mock up :-)
Is there any sane reason for this to be missing?

**Comment by JBZ-Fragmos** [comment]
honsetly i do not know reason why "31" is missing... maybe this list is defined in some original ISDA template in which for some reaons the list of permitted rollDays would stop at "30" but i'm not aware about it... 
but eventually, even if such ISDA documentation would exist, what say to a client actually trading and rolling 31 of the month ? can we just answer to the client "sorry, your actual booking data cannot be modelled/mapped in CDM because 31 is not a rollDay in CDM" ? currently i'm facing this issue for real... and i do not feel confortable about it...

**Comment by mgratacos** [comment]
What type of trade rolls on the 31st? What day count fraction do they use then? Just curious. I've never seen one. I thought 31st was never used, at least in the OTC market.

**Comment by JBZ-Fragmos** [comment]
trade at stake is a CCS made of 1M EUR leg vs. GBP-Sonia leg...

see below screenshot from Client Legacy system that Fragmos is mapping to CDM, hence the need for having this value available in CDM very quicly, please :

![image](https://github.com/finos/common-domain-model/assets/92978493/9c446462-6a23-49cf-a110-34e7797b105f)

on this extract from EUR leg, you can see for instance 1M Freq roll 1M Reset on the anchor day 31, floating rate is EURIBOR 1M tenor, DCF is Actual/360, businessConvention is ModFollowing, etc.

**Comment by mgratacos** [comment]
I am not an expert on this but I don't think it is correct to have a roll of 31st with a 1M frequency since you'll have months with 30 days. I think in this case firms use EOM (End of Month) and they roll on the 31st or 30th depending on the month (28th or 29th on Feb).

**Comment by harry-mcallister-bnpp** [comment]
RollConventionEnum in CDM derives from the related enumeration of the same name in FpML. The purpose of the enum is to control values produced to specifiy the "roll date" (or "roll day") on which successive periods of an expanded interest rate calculation schedule begin and end. "31" is not a value of the enumeration because not every month has 31 days (obvs). Consider a quarerly schedule commencing on December 31: then the successive roll dates are March 31, June **30**, September **30** (and December 31 again). It is therefore impossible - and inappropriate - in this case to specify the roll convention as a fixed day number; FpML provides the value "EOM" ("End Of Month") for this purpose instead.

A moment's thought suffices to show that there are no Quarterly schedules, and only one Semi-annual schedule (January/July) which could be specified using roll convention = "31". The contract example presented by _JBZ-Fragmos_ (above) has a _monthly_ reset schedule, in which case the sequence of roll dates is: {31, **28/29**, 31, **30**, 31, **30**, 31, 31, **30**, 31, **30**, 31}. Clearly the property _SCHED_Reset_AnnDay_ = "31" in the client's model is a _proxy_ for "EOM" roll dates and **must** be mapped accordingly in the CDM representation.

In summary, the case of roll day = "31" is supported by "EOM", by design. The Fragmos proposal should be rejected.

**Comment by JBZ-Fragmos** [comment]
thanks for the answer, that is helping, at least i can have short term solution to map the client value with an existing CDM value that is EOM if my understanding is correct

that being said, please still re-consider my proposal to add "31" in regards of the two "arguments" below : 

-  you mention above the fact _"31" is not a value of the enumeration because not every month has 31 days (obvs)"_ then why does "30" exists in the Enum list ? _obviously not every month has 30 days neither_... same as for 31 then...  following your comment, should our conclusion be to remove also "30" from the list, then ? I guess no... and then because 30 and 31 share same properties then let's also include 31...

- roll end of month without EOM abstract value might be appropriate : say you would roll 2 months basis e.g. 31 march, 31 May, 31 July, etc. though have not faced such business case, I see no conceptual issue to consider it, therefore why preventing users to represent something that is just possible per se ?

**Comment by harry-mcallister-bnpp** [comment]
- The purpose of including 30 (like all of the preceding integer values) is to support specification of a schedule which rolls consistently _on the 30th_ (as opposed to, say, the month-end date). By contrast, a schedule commencing on (or containing) month day 31 can _only_ be a EOM schedule.
- 31 was omitted from the FpML enumeration intentionally, to avoid duplicative ways of specifying the same outcome as EOM, and because the industry parlance speaks in terms of "end-of-month rolls" rather than (say) "31st rolls".
- The example 2 month schedule is not a plausible use case as there are no remaining 2 month term rates (and please note that continuing the sequence gives **30** September and **30** November as the next dates in the schedule). Users are not "prevented" from representing this schedule, as it remains consistent with end-of-month roll dates.
- It is important to note that there are number of roll conventions in common use which do not conform to a specific month day e.g. IMM (third Wednesday) dates, and their AUD/CAD/NZD variants. Any technical solution for schedule generation has to support these methods in addition to simple date arithmetic. In this context, EOM is just another "recipe" (albeit a simple one) for generating roll dates.

**Comment by JBZ-Fragmos** [comment]
thank you (therefore i'm closing this ticket)
